### PR TITLE
In app download

### PR DESCRIPTION
**Pull request by constraintAutomaton**
---
In app download
---

**Pull Request Type**
Please select what type of pull request this is:
- [ ] Bugfix
- [x] Feature Implementation

**Related issue**
related to #1772

**Description**
The first step to add a downloads tab. This feature changes the old mechanism to download media to one where the download is carry in app. From the settings menu, the user can define a working folder or demand to be asking every time where to download the media. The progress is displayed at intervals of around 20% into a toast.

**Screenshots**

**Testing**
ask for download path

- Go into option
- Click into the download setting
- set to ask for download path
- Click on the download button in a video page 
- Check if a save dialog appears
- Check if some toast displays the progress
- Check if the file is downloaded and functional

select  download path

- Go into option
- Click into the download setting
- Set to do not ask for download path
- click choose the path
- check if a directory dialog appears
- check if the path is written into the display
- Click on the download button in a video page 
- Check if some toast displays the progress
- Check if the file is downloaded and functional

try with multiple downloads links

**Desktop**
 - OS: Ubuntu
- OS Version: 20.04.3 LTS
- FreeTube version: v0.15.1 Beta

**Additional context**
The next step will be to create a download tab, the one after that is to enable the possibility to read the download file instead of streaming the media.

### CODE REVIEW

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Cant wait to test this out🎉🎉🎉

**Comment by bachig26** [comment]
does this implementation has the option to choose the quality (just for download) or does it downloads the selected watch quality?

**Comment by bachig26** [comment]
also, the notifications from the screenshots definitely look distracting too much, can it be made overlap over each other instead of the timed disappear or creating a text flow in the interface?

**Comment by constraintAutomaton** [comment]
> does this implementation has the option to choose the quality (just for download) or does it downloads the selected watch quality?

yes it's same as before you select the quality/type of file.

>  also, the notifications from the screenshots definitely look distracting too much, can it be made overlap over each other instead of the timed disappear or creating a text flow in the interface?

It only do that when it's a very small file (because the download will happen very fast) it sends a notification when the download start, by increase of ~20% and at the end of the download. I think I can create a mechanism to also based the frequency of notification around a time elapses, but I'm not sure if it should be a mechanism for the whole app instead of only for this feature if it is for the former than maybe #1049 should also solve this problem.

**Comment by bachig26** [comment]
> > does this implementation has the option to choose the quality (just for download) or does it downloads the selected watch quality?
> 
> It's same as before you select the quality/type of file from the download button.
> 
> > also, the notifications from the screenshots definitely look distracting too much, can it be made overlap over each other instead of the timed disappear or creating a text flow in the interface?
> 
> It only do that when it's a very small file (because the download will happen very fast) it sends a notification when the download start, by increase of ~20% (of the data received) and at the end of the download. I think I can create a mechanism to also based the frequency of notification around the time elapses, but I'm not sure if it should be a mechanism for the whole app or of only for this feature, if it is for the former than maybe #1049 should also solve this problem.

Thank you, that's good here.

regarding the notification, does it have a timer function/ mechanism to define how long it is visible?
also, does this pull has a snapshot? I haven't checked for its [snapshot] availability yet.

**Comment by constraintAutomaton** [comment]
> > > does this implementation has the option to choose the quality (just for download) or does it downloads the selected watch quality?
> > 
> > 
> > It's same as before you select the quality/type of file from the download button.
> > > also, the notifications from the screenshots definitely look distracting too much, can it be made overlap over each other instead of the timed disappear or creating a text flow in the interface?
> > 
> > 
> > It only do that when it's a very small file (because the download will happen very fast) it sends a notification when the download start, by increase of ~20% (of the data received) and at the end of the download. I think I can create a mechanism to also based the frequency of notification around the time elapses, but I'm not sure if it should be a mechanism for the whole app or of only for this feature, if it is for the former than maybe #1049 should also solve this problem.
> 
> Thank you, that's good here.
> 
> regarding the notification, does it have a timer function/ mechanism to define how long it is visible? also, does this pull has a snapshot? I haven't checked for its [snapshot] availability yet.

> also, does this pull has a snapshot? I haven't checked for its [snapshot] availability yet.

what do you mean by a snapshot? If you want to try it you can clone [my branch](https://github.com/constraintAutomaton/FreeTube/tree/feature/inAppDownload)

> regarding the notification, does it have a timer function/ mechanism to define how long it is visible?

no but both could be changed, I was more thinking in the line that if a 20% increase happen really fast to not show it. for example if it take 1sec to download just show the starting and ending message.

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — DISMISSED** [review]
LGTM! Everything works as intended. i really hate the in progress toasts but i think those will be replaced when there is a download tab with progress of all the videos ur downloading otherwise u'll receive a toast spam :(

**Comment by constraintAutomaton** [comment]
> LGTM! Everything works as intended. i really hate the in progress toasts but i think those will be replaced when there is a download tab with progress of all the videos ur downloading otherwise u'll receive a toast spam :(

that's my plan I want to solve the #1772 in two pr, but if it's really hated (2 people for the moment) I can keep the starting and ending toast or just have a toast at around 50% plus the starting and ending. I just don't like having no visual log at all.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
I was thinking about something like this for the download tab.

The only toast u'll need is The download has started and the download of video x is finished

**Comment by constraintAutomaton** [comment]
> 
> 
> I was thinking about something like this for the download tab.
> 
> The only toast u'll need is The download has started and the download of video x is finished

Yes, I was thinking about this too for a downloading page (for a next pr). I will change the toast.

File: src/renderer/store/modules/utils.js
Comment (efb4f5ff-1298-471a-8973-3d47447115dc) [inline src/renderer/store/modules/utils.js:182]: Downloading ${title} has finished/completed

Not sure which one is better

File: src/renderer/store/modules/utils.js
Comment (efb4f5ff-1298-471a-8973-3d47447115dc) [inline src/renderer/store/modules/utils.js:229]: Unable

File: src/renderer/store/modules/utils.js
Comment (efb4f5ff-1298-471a-8973-3d47447115dc) [inline src/renderer/store/modules/utils.js:185]: Downloading ${title} has started

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — COMMENTED** [review]
Some nitpicks

File: src/renderer/store/modules/utils.js
Comment (bachig26) [reply src/renderer/store/modules/utils.js:229]: "Unable to download ....." sounds good.

File: src/renderer/store/modules/utils.js
Comment (bachig26) [reply src/renderer/store/modules/utils.js:182]: For me, "Downloading ${title} has completed" sounds good.

File: src/renderer/store/modules/utils.js
Comment (ChunkyProgrammer) [inline src/renderer/store/modules/utils.js:215]:
```suggestion
        // mechanism kept for an upcoming download page
```

File: src/renderer/store/modules/utils.js
Comment (ChunkyProgrammer) [inline src/renderer/store/modules/utils.js:193]:
```suggestion
    //  mechanism to show the download progress reference https://javascript.info/fetch-progress
```

**Review by ChunkyProgrammer — COMMENTED** [review]
Maybe the toast messages should be translatable?

**Comment by bachig26** [comment]
I have doubt, is the download function/ mechanism used in the FreeTube based on yt-dlp or the original youtube-dl?

**Comment by constraintAutomaton** [comment]
> I have doubt, is the download function/ mechanism used in the FreeTube based on yt-dlp or the original youtube-dl?

it's [ytdl-core](https://www.npmjs.com/package/ytdl-core)

**Comment by bachig26** [comment]
> > I have doubt, is the download function/ mechanism used in the FreeTube based on yt-dlp or the original youtube-dl?
> 
> it's [ytdl-core](https://www.npmjs.com/package/ytdl-core)

so, does the module's update check done automatically, or has it been disabled for manual push?

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — DISMISSED** [review]
LGTM!

**Comment by constraintAutomaton** [comment]
> > > I have doubt, is the download function/ mechanism used in the FreeTube based on yt-dlp or the original youtube-dl?
> > 
> > 
> > it's [ytdl-core](https://www.npmjs.com/package/ytdl-core)
> 
> so, does the module's update check done automatically, or has it been disabled for manual push?

I'm not sure if I understand if it's about reetube updating the package to keep up with release I'm not sure but right now there is a [bot updating node-forge](https://github.com/FreeTubeApp/FreeTube/pull/1987) so it might be the case but I don't know.

**Comment by ChunkyProgrammer** [comment]
Dependabot is only configured for security updates. Other updates are manual

File: src/renderer/components/download-settings/download-settings.sass
Comment (PikachuEXE) [inline src/renderer/components/download-settings/download-settings.sass:8]: I know this is a probably minor issue but this project does have setting to ensure final new line

https://github.com/FreeTubeApp/FreeTube/blob/development/.editorconfig#L9

File: src/renderer/store/modules/utils.js
Comment (PikachuEXE) [inline src/renderer/store/modules/utils.js:178]: I think passing an object is better than an array (as you see for `{ rootState, dispatch }`)
```suggestion
  async downloadMedia({ rootState, dispatch }, { url, title, extension, folder }) {
```

File: src/renderer/components/ft-icon-button/ft-icon-button.js
Comment (PikachuEXE) [inline src/renderer/components/ft-icon-button/ft-icon-button.js:131]:
```suggestion
      this.$emit('click', {
        url: this.dropdownValues[index], 
        title: this.relatedVideoTitle,
        extension: this.fileExtension[index],
        folder: this.$store.getters.getDownloadFolder
      })
```

**Review by PikachuEXE — CHANGES_REQUESTED** [review]
Mainly use "named parameters" style

**Comment by peepopoggers** [comment]
toasts like that are annoying if too often :<

**Comment by constraintAutomaton** [comment]
> toasts like that are annoying if too often :<

Right now there is only the starting and finishing toast.

**Comment by bachig26** [comment]
@constraintAutomaton Hi, Have you checked with the invidious fallback videos for download?
due to current youtube's 18+ restriction for some videos, so will it work for that videos (invidious fallback)?

**Comment by constraintAutomaton** [comment]
> @constraintAutomaton Hi, Have you checked with the invidious fallback videos for download? due to current youtube's 18+ restriction for some videos, so will it work for that videos (invidious fallback download)?

I've tested it and it works anyway, the mechanism by which the app obtains the video files has not changed only the way the user downloads it.

File: src/renderer/store/modules/settings.js
Comment (PikachuEXE) [inline src/renderer/store/modules/settings.js:220]:
```suggestion
  downloadFolderPath: ''
```

File: src/renderer/components/download-settings/download-settings.js
Comment (PikachuEXE) [inline src/renderer/components/download-settings/download-settings.js:20]:
```suggestion
      askForDownloadPath: this.$store.getters.getDownloadFolderPath === ''
```

File: src/renderer/components/download-settings/download-settings.js
Comment (PikachuEXE) [inline src/renderer/components/download-settings/download-settings.js:25]:
```suggestion
      return this.$store.getters.getDownloadFolderPath
```

File: src/renderer/components/download-settings/download-settings.js
Comment (PikachuEXE) [inline src/renderer/components/download-settings/download-settings.js:32]:
```suggestion
        this.updateDownloadFolderPath('')
```

File: src/renderer/components/download-settings/download-settings.js
Comment (PikachuEXE) [inline src/renderer/components/download-settings/download-settings.js:42]:
```suggestion
      this.updateDownloadFolderPath(folder.filePaths[0])
```

File: src/renderer/components/download-settings/download-settings.js
Comment (PikachuEXE) [inline src/renderer/components/download-settings/download-settings.js:45]:
```suggestion
      'updateDownloadFolderPath'
```

File: src/renderer/components/ft-icon-button/ft-icon-button.js
Comment (PikachuEXE) [inline src/renderer/components/ft-icon-button/ft-icon-button.js:135]:
```suggestion
        folderPath: this.$store.getters.getDownloadFolderPath
```

File: src/renderer/store/modules/utils.js
Comment (PikachuEXE) [inline src/renderer/store/modules/utils.js:178]:
```suggestion
  async downloadMedia({ rootState, dispatch }, { url, title, extension, folderPath }) {
```

File: src/renderer/store/modules/utils.js
Comment (PikachuEXE) [inline src/renderer/store/modules/utils.js:180]:
```suggestion
    const askFolderPath = folderPath === ''
```

File: src/renderer/store/modules/utils.js
Comment (PikachuEXE) [inline src/renderer/store/modules/utils.js:188]:
```suggestion
    if (askFolderPath) {
```

File: src/renderer/store/modules/utils.js
Comment (PikachuEXE) [inline src/renderer/store/modules/utils.js:237]:
```suggestion
    if (usingElectron && !askFolderPath) {
```

File: src/renderer/store/modules/utils.js
Comment (PikachuEXE) [inline src/renderer/store/modules/utils.js:239]:
```suggestion
      fs.writeFile(`${folderPath}/${title}.${extension}`, new DataView(buffer), (err) => {
```

**Review by PikachuEXE — CHANGES_REQUESTED** [review]
I think the variables with suffix `folder` storing folder paths should be renamed
Without renaming them there might be unnecessary confusion 
(except `folder.filePaths[0]` is named correctly)

**Review by PikachuEXE — DISMISSED** [review]

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Is it possible to show the download has started toast after u select a path? Toast is now hidden by popup window for choosing download path and if u click cancel on the popup window u still have seen that toast even though the download hasn't started

https://user-images.githubusercontent.com/73130443/149418993-d1408c93-d8fd-462d-9a99-28f25ad83e40.mp4

**Comment by constraintAutomaton** [comment]
> Is it possible to show the download has started toast after u select a path? Toast is now hidden by popup window for choosing download path and if u click cancel on the popup window u still have seen that toast even though the download hasn't started
> VirtualBoxVM_WnyiE78jDu.mp4
> 
> edit: did u somehow add this icon into the mediaplayer?
> 
> Capture

I have no idea how this icon apear

**Comment by constraintAutomaton** [comment]
> Is it possible to show the download has started toast after u select a path? Toast is now hidden by popup window for choosing download path and if u click cancel on the popup window u still have seen that toast even though the download hasn't started
> VirtualBoxVM_WnyiE78jDu.mp4
> 
> edit: did u somehow add this icon into the mediaplayer?
> 
> Capture

I don't know how it happen to be "activated" but I deleted it from the control bar

**Review by PikachuEXE — DISMISSED** [review]

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — DISMISSED** [review]

**Comment by bachig26** [comment]
> Is it possible to show the download has started toast after u select a path? Toast is now hidden by popup window for choosing download path and if u click cancel on the popup window u still have seen that toast even though the download hasn't started
> 
>  VirtualBoxVM_WnyiE78jDu.mp4 
> edit: did u somehow add this icon into the mediaplayer?
> 
> Capture

@constraintAutomaton from this video clip, I see that after selecting the quality, it asks for the "**download folder path**". 
But if the "**download folder path**" is already selected in the download settings, then it doesn't ask again, right?

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — DISMISSED** [review]
LGTM!

**Comment by constraintAutomaton** [comment]
> > Is it possible to show the download has started toast after u select a path? Toast is now hidden by popup window for choosing download path and if u click cancel on the popup window u still have seen that toast even though the download hasn't started
> > VirtualBoxVM_WnyiE78jDu.mp4
> > edit: did u somehow add this icon into the mediaplayer?
> > Capture
> 
> @constraintAutomaton @efb4f5ff-1298-471a-8973-3d47447115dc from this video clip, I see that after selecting the quality, it asks for the "**download folder path**". But if the "**download folder path**" is already selected in the download settings, then it doesn't ask again, right?

There are two modes, the first one you select a downloading repertory where every time you do a download operation it will be saved inside this repertory. The second is where you don't select a downloading repertory in that case EACH time you download media, it will ask for a directory. So it much like what web browser tend to do (Firefox, chrome).

**Comment by PikachuEXE** [comment]
I have tested 2nd mode: "where every time you do a download operation it will be saved inside this folder"

And it does not handle missing folder gracefully

image

**Comment by constraintAutomaton** [comment]
> I have tested 2nd mode: "where every time you do a download operation it will be saved inside this folder"
> 
> And it does not handle missing folder gracefully
> 
> image

I'm not sure I understand what it should do if the directory doesn't exist, the user can't define a non existing folder in either mode (unless it's a bug I didn't saw or the user delete it afterward) and what do you mean by gratefully, should it create the directory and the file or the error label should be different?

**Comment by constraintAutomaton** [comment]
> I have tested 2nd mode: "where every time you do a download operation it will be saved inside this folder"
> 
> And it does not handle missing folder gracefully
> 
> image

I've changed the error message and implemented a falling back i the folder doesn't exist anymore

**Comment by bachig26** [comment]
> > I have tested 2nd mode: "where every time you do a download operation it will be saved inside this folder"
> > And it does not handle missing folder gracefully
> > image
> 
> I've changed the error message and implemented a falling back i the folder doesn't exist anymore

is it possible to define a function to create the folder if it doesn't exist?

File: static/locales/en-US.yaml
Comment (bachig26) [inline static/locales/en-US.yaml:718]: I think this one has a misspelling, "**repertory**" which I believe you wanted to give it as "**repository**" but for this position "**directory**" seems to be the correct one here.

**Comment by PikachuEXE** [comment]
When the folder indicated by **previously set** folder path does not exist
It's an **exception** (unexpected by program at least)
Since it's **previously set** by user, this event should also be unexpected by the user.
And thus we should ask for user action, ask for another path or proper error message.

Will test the fallback soon

**Review by PikachuEXE — APPROVED** [review]
Tested fallback mode
Though the UX is not perfect in such case but acceptable (and I have no idea for improvement right now)

**Comment by constraintAutomaton** [comment]
Hi, I think I've made all the change requested so far I would appreciate new reviews.

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — APPROVED** [review]
LGTM!

**Comment by PikachuEXE** [comment]
Get one more person to review this?

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Everyone is very busy i think the approval process really slowed down :(

**Comment by bachig26** [comment]
> Everyone is very busy i think. the approval process really slowed down :(

I think the number of reviewers can be reduced to 2 if possible temporarily.

**Comment by PikachuEXE** [comment]
But 2 is too easy to pass hmmmm

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Yeah 2 is to easy in some cases the third reviewer find things that the other two overlooked

**Comment by constraintAutomaton** [comment]
maybe a tag to show when pr is waiting for review since a long time would be nice. There are some old prs that are still waiting for reviews whereas newer pr have been merged.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
The labels are doing the job for us.

Before we took a break in December we merged allot of pr's with help of the labeling system.

But now almost everyone within the team is very busy and thats why progress is so slow

**Comment by PikachuEXE** [comment]
> But now almost everyone within the team is very busy and thats why progress is so slow

The key point is we need 3 to merge but only got 2 active a.t.m. 😆

**Review by ChunkyProgrammer — APPROVED** [review]
A few things I've noticed but I'm okay with merging this PR now.

**Comment by bachig26** [comment]
@PrestonN can we get a executable file with this PR?

**Comment by ChunkyProgrammer** [comment]
You're able to download the nightly build from here: https://github.com/FreeTubeApp/FreeTube/actions/runs/1769337129

**Comment by constraintAutomaton** [comment]
> A few things I've noticed but I'm okay with merging this PR now.

I wouldn't have mind to fix it before the merge. I'm working on a download tab and I will fix it there.

**Comment by PrestonN** [comment]
Hey there! I was a little too late to getting to this, but there were a couple of small issues that needed to be changed with this PR.

- The French translation changed you provided seemed to have a few issues with being parsed. It specifically seemed to be because of words that used single quotes while the string itself was also being initialized with single quotes.
- The card for download settings should be moved to be above Sponsorblock. I specifically want to keep third party services / implementations outside of Invidious as last on the list of settings. This is mostly a personal preference and not something that I've publicly made clear, so this isn't your fault.

I've gone ahead and fixed both of the issues above, this is mostly just me making you aware of the changes that was made. Thank you very much for the PR and hopefully we'll see more from you in the future. :)

**Comment by PrestonN** [comment]
Hi again!

After spending a little bit of time in the dev environment, I've noticed a couple of more small things about this PR that seems to break things / seems odd. I wanted to mention them here for future reference.

- The "Download Started" message still shows up if I cancel the download. It's an easy fix, but still can be confusing for others that would be using this feature.
- This PR breaks functionality with other places that uses `ft-icon-button` that is not the download button. Namely the video format button and the upcoming playlist buttons I've added in a separate branch. The `ft-icon-button` component is meant to be very basic / generic and adding functionality to that component for one specific feature isn't recommended since it's used in a variety of places. I would have much rather seen `ft-icon-button` not be touched at all and have the core logic moved over to `watch-video-info`.
- I _really_ don't like using `while (true)` for any logic. There's very few situations where that would be the best approach for implementing features and is bound to cause issues in the future. I haven't looked too much into it yet but it's possible that `downloadMedia` can be moved to a component as well.

I apologize in advance if I end up sounding too harsh when I say this, but this PR likely should have been reviewed a bit more before being merged. I would likely have reverted the changes in dev to give this PR some more time, however several changes have been made since this has been merged. I will likely be reworking a few things from this PR myself here in the next day or so so that we can prepare ourselves for a new release. I'm still grateful for the effort you've put into this PR, but I want to make sure that additions to the code are sound and setup / located in ways that make the most sense. Again, I'm very sorry if this sounds too harsh, I don't want it to seem that way.

**Comment by constraintAutomaton** [comment]
> Hi again!
> 
> After spending a little bit of time in the dev environment, I've noticed a couple of more small things about this PR that seems to break things / seems odd. I wanted to mention them here for future reference.
> 
>     * The "Download Started" message still shows up if I cancel the download. It's an easy fix, but still can be confusing for others that would be using this feature.
> 
>     * This PR breaks functionality with other places that uses `ft-icon-button` that is not the download button. Namely the video format button and the upcoming playlist buttons I've added in a separate branch. The `ft-icon-button` component is meant to be very basic / generic and adding functionality to that component for one specific feature isn't recommended since it's used in a variety of places. I would have much rather seen `ft-icon-button` not be touched at all and have the core logic moved over to `watch-video-info`.
> 
>     * I _really_ don't like using `while (true)` for any logic. There's very few situations where that would be the best approach for implementing features and is bound to cause issues in the future. I haven't looked too much into it yet but it's possible that `downloadMedia` can be moved to a component as well.
> 
> 
> I apologize in advance if I end up sounding too harsh when I say this, but this PR likely should have been reviewed a bit more before being merged. I would likely have reverted the changes in dev to give this PR some more time, however several changes have been made since this has been merged. I will likely be reworking a few things from this PR myself here in the next day or so so that we can prepare ourselves for a new release. I'm still grateful for the effort you've put into this PR, but I want to make sure that additions to the code are sound and setup / located in ways that make the most sense. Again, I'm very sorry if this sounds too harsh, I don't want it to seem that way.

I didn't notices those issue and I wouldn't have mind to work on them before the merge. I also have some free time and I can fix them/some of them if needed.

**Comment by AshishSPu** [comment]
Hey there!

Since I found no other place to drop my query, I will do so here,

Is there any way to access the progress bar for downloads? 

I had kept a video for download and suddenly the download stopped (as evident from the network speed in Task Manager) after an hour but the video does not appear in my Files Explorer. Sorry if this is not the right place for this.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
This is not possible yet. The next implementation will include a download tab with all the info u need like download progress

**Comment by AshishSPu** [comment]
Okay! Will be eagerly waiting for that implementation then. Meanwhile, I would rely on Newpipe on phone for downloading contents. 

Thank you for responding.

**Pull request merged**
